File: android/measure/src/main/java/sh/measure/android/tracing/MsrSpanBuilder.kt
package sh.measure.android.tracing

import sh.measure.android.SessionManager
import sh.measure.android.logger.Logger
import sh.measure.android.utils.IdProvider
import sh.measure.android.utils.Sampler
import sh.measure.android.utils.TimeProvider

internal class MsrSpanBuilder(
    val name: String,
    private val idProvider: IdProvider,
    private val timeProvider: TimeProvider,
    private val spanProcessor: SpanProcessor,
    private val sessionManager: SessionManager,
    private val sampler: Sampler,
    private val logger: Logger,
) : SpanBuilder {
    private var parentSpan: Span? = null

    override fun setParent(span: Span): SpanBuilder {
        this.parentSpan = span
        return this
    }

    override fun startSpan(): Span = MsrSpan.startSpan(
        name = name,
        logger = logger,
        timeProvider = timeProvider,
        spanProcessor = spanProcessor,
        sessionManager = sessionManager,
        idProvider = idProvider,
        parentSpan = parentSpan,
        sampler = sampler,
    )

    override fun startSpan(timeMs: Long): Span {
        val span = MsrSpan.startSpan(
            name = name,
            logger = logger,
            timeProvider = timeProvider,
            idProvider = idProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            parentSpan = parentSpan,
            timestamp = timeMs,
            sampler = sampler,
        )
        return span
    }
}

File: flutter/packages/measure_flutter/lib/src/tracing/msr_span_builder.dart
import 'package:measure_flutter/src/tracing/span.dart';
import 'package:measure_flutter/src/tracing/span_builder.dart';
import 'package:measure_flutter/src/tracing/span_processor.dart';
import 'package:measure_flutter/src/utils/sampler.dart';

import '../logger/logger.dart';
import '../time/time_provider.dart';
import '../utils/id_provider.dart';
import 'msr_span.dart';

class MsrSpanBuilder extends SpanBuilder {
  final String name;
  final IdProvider _idProvider;
  final TimeProvider _timeProvider;
  final SpanProcessor _spanProcessor;
  final Sampler _sampler;
  final Logger _logger;

  Span? _parentSpan;

  MsrSpanBuilder({
    required this.name,
    required IdProvider idProvider,
    required TimeProvider timeProvider,
    required SpanProcessor spanProcessor,
    required Sampler sampler,
    required Logger logger,
  })  : _idProvider = idProvider,
        _timeProvider = timeProvider,
        _spanProcessor = spanProcessor,
        _sampler = sampler,
        _logger = logger;

  @override
  SpanBuilder setParent(Span span) {
    _parentSpan = span;
    return this;
  }

  @override
  Span startSpan({int? timestamp}) {
    return MsrSpan.startSpan(
      name: name,
      logger: _logger,
      timeProvider: _timeProvider,
      spanProcessor: _spanProcessor,
      idProvider: _idProvider,
      parentSpan: _parentSpan,
      sampler: _sampler,
      timestamp: timestamp,
    );
  }
}
